### PR TITLE
add kernel coredump and analysis on sonic kernel

### PR DESCRIPTION
**Pull request by sun-siyuan**
Summary: add kdump package to j2 template and config after

this is the to add kdump to capture the kernel crash core and for further analysis by crash tool

in this PR, contain two part

1,  install kdump tool chain to host environment
2,  configure kdump tool in both boot up via grub.cfg and system level

test done:

test build process, build sonic-broadcom.bin and sonic-aboot-broadcom.swi

-rwxr-xr-x  1 sun sun 562086493 Jul 28 21:28 sonic-aboot-broadcom.swi
-rw-r--r--  1 sun sun    215882 Jul 28 21:28 sonic-aboot-broadcom.swi.log
-rwxr-xr-x  1 sun sun 569867542 Jul 28 00:52 sonic-broadcom.bin
-rw-r--r--  1 sun sun    285585 Jul 28 00:52 sonic-broadcom.bin.log

test image, load sonic-broadcom.bin to switch

 Signed-off-by: siyuan sun siyuan.sun@alibaba-inc.com

### CODE REVIEW

**Comment by lguohan** [comment]
can you describe the test you have done?

File: files/image_config/platform/rc.local [hunk 1]
```diff
@@ -318,6 +318,11 @@ if [ -f $FIRST_BOOT_FILE ]; then
         # Initialize the SONiC's grub config
         mv /host/grub.cfg /host/grub/grub.cfg
     fi
+    sed -i 's/[^M] quiet/ crashkernel=768M quiet/' /host/grub/grub.cfg
```
Comment (jipanyang) [inline files/image_config/platform/rc.local]: Please check whether 768M is absolutely necessary, 256M could be more reasonable?

**Comment by sun-siyuan** [comment]
yes, 256M is correct,  I cherry-pick from dev-master, let me check why this happened


------------------------------------------------------------------

**Comment by pavel-shirshov** [comment]
Can you please share console output after you issued commands
```
echo 1 > /proc/sys/kernel/sysrq
echo c > /proc/sysrq-trigger
```

Thanks

**Comment by sun-siyuan** [comment]
> Can you please share console output after you issued commands
> 
> ```
> echo 1 > /proc/sys/kernel/sysrq
> echo c > /proc/sysrq-trigger
> ```
> 
> Thanks

sure, please see below:


[1568872.292681] sysrq: SysRq : Trigger a crash
[1568872.360135] BUG: unable to handle kernel NULL pointer dereference at           (null)
[1568872.361626] IP: [<ffffffffba828802>] sysrq_handle_crash+0x12/0x20
[1568872.383586] PGD 8000000154beb067 [1568872.384199] PUD 154b30067
PMD 0 [1568872.384888]
[1568872.385263] Oops: 0002 [#1] SMP
[1568872.385880] Modules linked in: fuse veth dummy iptable_raw xt_limit xt_tcpudp xt_conntrack bridge stp llc nf_conntrack_netlink nfnetlink xfrm_user xfrm_algo bf_tun(O) crct10dif_pclmul crc32_pclmul ghash_clmulni_intel joydev virtio_balloon virtio_console evdev ip6table_filter serio_raw button ip6_tables iptable_filter iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 nf_nat nf_conntrack pcspkr iptable_mangle ip_tables x_tables autofs4 loop ext4 crc16 jbd2 crc32c_generic fscrypto ecb mbcache nls_utf8 nls_cp437 nls_ascii vfat fat overlay squashfs ata_generic virtio_blk virtio_net crc32c_intel aesni_intel aes_x86_64 glue_helper lrw gf128mul ablk_helper cryptd psmouse ata_piix ehci_pci uhci_hcd libata ehci_hcd usbcore usb_common virtio_pci scsi_mod virtio_ring virtio i2c_piix4 floppy
[1568872.400602] CPU: 0 PID: 30349 Comm: bash Tainted: G           O    4.9.0-8-2-amd64 #1 Debian 4.9.110-3+deb9u6
[1568872.402364] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.10.2-1ubuntu1 04/01/2014
[1568872.411005] task: ffff97f79a67c000 task.stack: ffffb02508d3c000
[1568872.412108] RIP: 0010:[<ffffffffba828802>]  [<ffffffffba828802>] sysrq_handle_crash+0x12/0x20
[1568872.413698] RSP: 0018:ffffb02508d3fe78  EFLAGS: 00010282
[1568872.414682] RAX: ffffffffba8287f0 RBX: 0000000000000063 RCX: 0000000000000000
[1568872.415993] RDX: 0000000000000000 RSI: ffff97f87fc10648 RDI: 0000000000000063
[1568872.417315] RBP: ffffffffbb0bf2e0 R08: 0000000000000001 R09: 0000000000009868
[1568872.418630] R10: 0000000000000001 R11: 0000000000000001 R12: 0000000000000004
[1568872.419951] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[1568872.421274] FS:  00007f30474ee700(0000) GS:ffff97f87fc00000(0000) knlGS:0000000000000000
[1568872.422753] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[1568872.423824] CR2: 0000000000000000 CR3: 000000014fc6c000 CR4: 0000000000160670
[1568872.425148] Stack:
[1568872.425571]  ffffffffba828f37 0000000000000002 fffffffffffffffb ffffb02508d3ff08
[1568872.427071]  0000000001d1f008 ffffffffba82937b ffff97f86fad03c0 ffffffffba677b00
[1568872.428574]  0000000000000002 ffff97f874433200 ffffffffba608490 ffff97f874433200
[1568872.433765] Call Trace:
[1568872.434268]  [<ffffffffba828f37>] ? __handle_sysrq+0xf7/0x150
[1568872.435332]  [<ffffffffba82937b>] ? write_sysrq_trigger+0x2b/0x30
[1568872.450405]  [<ffffffffba677b00>] ? proc_reg_write+0x40/0x70
[1568872.451539]  [<ffffffffba608490>] ? vfs_write+0xb0/0x190
[1568872.452532]  [<ffffffffba6098d2>] ? SyS_write+0x52/0xc0
[1568872.453722]  [<ffffffffba403b7d>] ? do_syscall_64+0x8d/0xf0
[1568872.457141]  [<ffffffffbaa15e8e>] ? entry_SYSCALL_64_after_swapgs+0x58/0xc6
[1568872.458409] Code: 41 5c 41 5d 41 5e 41 5f e9 bc 1f cf ff 66 2e 0f 1f 84 00 00 00 00 00 66 90 0f 1f 44 00 00 c7 05 89 94 a8 00 01 00 00 00 0f ae f8 <c6> 04 25 00 00 00 00 01 c3 0f 1f 44 00 00 0f 1f 44 00 00 53 8d
[1568872.464225] RIP  [<ffffffffba828802>] sysrq_handle_crash+0x12/0x20
[1568872.465409]  RSP <ffffb02508d3fe78>
[1568872.466083] CR2: 0000000000000000
[    0.000000] do_IRQ: 0.113 No irq handler for vector
[    4.321878] kdump-tools[323]: Starting kdump-tools: running makedumpfile -c -d 31 /proc/vmcore /var/crash/201908022135/dump-incomplete.
Copying data                       : [100.0 %] \
[   22.021250] kdump-tools[323]: The kernel version is not supported.
[   22.023335] kdump-tools[323]: The makedumpfile operation may be incomplete.
[   22.025515] kdump-tools[323]: The dumpfile is saved to /var/crash/201908022135/dump-incomplete.
[   22.032202] kdump-tools[323]: makedumpfile Completed.
[   22.036659] kdump-tools[323]: kdump-tools: saved vmcore in /var/crash/201908022135.
[   22.432118] kdump-tools[323]: running makedumpfile --dump-dmesg /proc/vmcore /var/crash/201908022135/dmesg.201908022135.
[   22.436076] kdump-tools[323]: The kernel version is not supported.
[   22.440348] kdump-tools[323]: The makedumpfile operation may be incomplete.
[   22.443098] kdump-tools[323]: The dmesg log is saved to /var/crash/201908022135/dmesg.201908022135.
[   22.445600] kdump-tools[323]: makedumpfile Completed.
[   22.447797] kdump-tools[323]: kdump-tools: saved dmesg content in /var/crash/201908022135.
[   22.498068] kdump-tools[323]: Fri, 02 Aug 2019 21:35:39 +0000
[   22.511355] kdump-tools[323]: logname: no login name

**Comment by pavel-shirshov** [comment]
@sun-siyuan 
Thank you for the output.
But after the crash the system rebooted itself?

Thanks

**Comment by sun-siyuan** [comment]
yes, reboot  is the default action after crash. if we need to keep the data flow running temporarily without reboot, which could be done as well.


------------------------------------------------------------------

**Comment by lguohan** [comment]
how do we decide to reserve 256M?

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/kernel_administration_guide/kernel_crash_dump_guide

I also agree with pavel, we should make it optional.

**Comment by sun-siyuan** [comment]
> how do we decide to reserve 256M?
> 
> https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/kernel_administration_guide/kernel_crash_dump_guide
> 
> I also agree with pavel, we should make it optional.

we were using 768M according to redhat recommend but consider it could be a waste of running memory, and test with 256M which works fine.

I just update the PR to make it optional

**Comment by sun-siyuan** [comment]
> Can you please make this feature optional with default no?
> See comments from @lguohan

please check the new diff, which make this optional and default NO

**Comment by sun-siyuan** [comment]
looks like check failed due to below reason, not sure whether related to my change:

finished. No test results found.'
Using context: baseimage
java.lang.OutOfMemoryError: Java heap space
Finished: FAILURE

**Comment by sun-siyuan** [comment]
@pavel-shirshov please let me know whether you are ok with the new changes

**Comment by pavel-shirshov** [comment]
retest this please

File: files/image_config/platform/rc.local [hunk 2]
```diff
@@ -319,6 +319,15 @@ if [ -f $FIRST_BOOT_FILE ]; then
         mv /host/grub.cfg /host/grub/grub.cfg
     fi
 
+    if `dpkg -l kdump-tools > /dev/null 2>&1`;
+    then
+	sed -i 's/[^M] quiet/ crashkernel=256M quiet/' /host/grub/grub.cfg
+
+	kdump-config unload
+	kdump-config symlinks `uname -r`
+	kdump-config load
```
Comment (lguohan) [inline files/image_config/platform/rc.local]: this only works for the next reboot. can we get a solution that works for current boot?

Comment (sun-siyuan) [reply files/image_config/platform/rc.local]: agree but this is an known issue, it is hard to get it work with first  reboot due to the nature of the grub, and changes are made to the new image,while the sonic_installer used in the upgrade actually from the previous image.

**Comment by sun-siyuan** [comment]
> retest this please

keep hitting this error when rebuild, thought it maybe a glitch, but last several days even after pull rebase, any clue please?

# Obtaining the Debian kernel source
rm -rf linux-4.9.168
wget -O linux_4.9.168-1+deb9u3.dsc "http://security.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168-1+deb9u3.dsc"
wget -O linux_4.9.168.orig.tar.xz "http://security.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168.orig.tar.xz"
wget -O linux_4.9.168-1+deb9u3.debian.tar.xz "http://security.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168-1+deb9u3.debian.tar.xz"
dpkg-source -x linux_4.9.168-1+deb9u3.dsc
pushd linux-4.9.168
git init
git add -f *
git commit -qm "check in all loose files and diffs"
# patching anything that could affect following configuration generation.
stg init
stg import -s ../patch/preconfig/series
# re-generate debian/rules.gen, requires kernel-wedge
debian/bin/gencontrol.py
# generate linux build file for amd64_none_amd64
fakeroot make -f debian/rules.gen setup_amd64_none_amd64
# Applying patches and configuration changes
git add debian/build/build_amd64_none_amd64/.config -f
git add debian/config.defines.dump -f
git add debian/control -f
git add debian/rules.gen -f
git add debian/tests/control -f
git commit -m "unmodified debian source"
# Learning new git repo head (above commit) by calling stg repair.
stg repair
stg import -s ../patch/series
# Building a custom kernel from Debian kernel source
DO_DOCS=False fakeroot make -f debian/rules -j 24 binary-indep
fakeroot make -f debian/rules.gen -j 24 binary-arch_amd64_none
popd
mv linux-headers-4.9.0-9-2-amd64_4.9.168-1+deb9u3_amd64.deb linux-image-4.9.0-9-2-amd64_4.9.168-1+deb9u3_amd64.deb linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb /sonic/target/debs/stretch/
--2019-08-14 17:36:56--  http://security.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168-1+deb9u3.dsc
Resolving security.debian.org (security.debian.org)... 128.31.0.63, 128.101.240.215, 149.20.4.14, ...
Connecting to security.debian.org (security.debian.org)|128.31.0.63|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: http://security-cdn.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168-1+deb9u3.dsc [following]
--2019-08-14 17:36:57--  http://security-cdn.debian.org/debian-security/pool/updates/main/l/linux/linux_4.9.168-1+deb9u3.dsc
Resolving security-cdn.debian.org (security-cdn.debian.org)... 151.101.196.204, 2a04:4e42:6::204
Connecting to security-cdn.debian.org (security-cdn.debian.org)|151.101.196.204|:80... connected.
HTTP request sent, awaiting response... 404 Not Found
2019-08-14 17:36:57 ERROR 404: Not Found.

Makefile:56: recipe for target '/sonic/target/debs/stretch/linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb' failed
make[1]: *** [/sonic/target/debs/stretch/linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb] Error 8
make[1]: Leaving directory '/sonic/src/sonic-linux-kernel'
[  FAIL LOG END  ] [ target/debs/stretch/linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb ]
slave.mk:293: recipe for target 'target/debs/stretch/linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb' failed
make: *** [target/debs/stretch/linux-headers-4.9.0-9-2-common_4.9.168-1+deb9u3_all.deb] Error 1
Makefile.work:148: recipe for target 'target/sonic-broadcom.bin' failed
make[1]: *** [target/sonic-broadcom.bin] Error 2
make[1]: Leaving directory '/home/shadow/sun/ali/sonic/community/sonic-buildimage'
Makefile:6: recipe for target 'target/sonic-broadcom.bin' failed
make: *** [target/sonic-broadcom.bin] Error 2
mail: Null message body; hope that's ok
You have new mail in /var/mail/sun

File: files/image_config/platform/rc.local (hunk 2 above)
Comment (lguohan) [reply files/image_config/platform/rc.local]: you can modify the installer/x86_64/install.sh file

File: files/build_templates/sonic_debian_extension.j2 [hunk 3]
```diff
@@ -331,6 +331,17 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get clean -y
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get autoremove -y
 
+# install kernel dump utility
+{%- if sonic_kdump_enable == "y" %}
+{% set coredump = [ "crash", "makedumpfile" ] -%}
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: crash is analyze tools, it should only be installed in the dbg image. not in normal image. In normal image, you can only install kdump-tools. makedumpfile will be installed when you install kdump-tools

File: files/build_templates/sonic_debian_extension.j2 [hunk 4]
```diff
@@ -331,6 +331,17 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get clean -y
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get autoremove -y
 
+# install kernel dump utility
+{%- if sonic_kdump_enable == "y" %}
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: it is only to install kdump-tools by default, only the kernel option should be configurable.

File: files/build_templates/sonic_debian_extension.j2 [hunk 5]
```diff
@@ -331,6 +331,17 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
...
+{% set coredump = [ "crash", "makedumpfile" ] -%}
+{% for pkg in coredump -%}
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get install -y {{ pkg }}
+{% endfor -%}
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get install -y kdump-tools || true
+sudo sed -i 's/\/MODULES=dep\//\/MODULES=most\//' $FILESYSTEM_ROOT/etc/kernel/postinst.d/kdump-tools
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: why do you need most? which filesystem do you need?

**Comment by lguohan** [comment]
besides the build option, I think we should have a command line to enable/disable this kernel crash dump feature.

like

config kdump enable --size=265M
config kdump disable

**Comment by lguohan** [comment]
is there any plan to update the PR based on the feedback?

File: files/image_config/platform/rc.local (hunk 2 above)
Comment (sun-siyuan) [reply files/image_config/platform/rc.local]: thanks for pointing out this , I move grub.cfg change to install.sh and make it optional depends on build option

File: files/build_templates/sonic_debian_extension.j2 (hunk 5 above)
Comment (lguohan) [reply files/build_templates/sonic_debian_extension.j2]: can you address this question?

**Comment by sun-siyuan** [comment]
this is addressed as well, it will be updated once I get all test done


------------------------------------------------------------------

File: files/build_templates/sonic_debian_extension.j2 (hunk 5 above)
Comment (sun-siyuan) [reply files/build_templates/sonic_debian_extension.j2]: kdump-tools hit a post-installation issue, I have work around it this way

**Comment by sun-siyuan** [comment]
retest due to java heap overflow

**Comment by sun-siyuan** [comment]
retest

**Comment by sun-siyuan** [comment]
retest this please

**Comment by sun-siyuan** [comment]
retest this please

**Comment by sun-siyuan** [comment]
@pavel-shirshov

mellanox build server run out of space, fail twice when build, please fix it so I can continue, thanks

**
gzip: stdout: No space left on device
slave.mk:521: recipe for target 'target/docker-database.gz' failed
make: *** [target/docker-database.gz] Error 1
Makefile.work:191: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/data/johnar/workspace/mellanox/buildimage-mlnx-all-pr'
Makefile:6: recipe for target 'all' failed
make: *** [all] Error 2
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Declarative: Post Actions)
[Pipeline] cleanWs
[WS-CLEANUP] Deleting project workspace...
[WS-CLEANUP] Deferred wipeout is used...
[WS-CLEANUP] done
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Adding one-line test results to commit status...
Setting status of 4a683b0e0ac5c8adb4c693248cd1cd970a4419c2 to FAILURE with url https://sonic-jenkins.westus2.cloudapp.azure.com/job/mellanox/job/buildimage-mlnx-all-pr/890/ and message: 'Build finished. No test results found.'
Using context: mellanox
ERROR: script returned exit code 2
Finished: FAILURE**

**Comment by lguohan** [comment]
retest Mellanox please

**Comment by sun-siyuan** [comment]
@lguohan
thanks for helping retest mellanox, seems all checks passed now, I have test the build with ansible 100 times as well. please help to close this PR.

File: files/build_templates/sonic_debian_extension.j2 [hunk 6]
```diff
@@ -353,6 +353,14 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get clean -y
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get autoremove -y
 
+# install kernel dump utility
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get install -y kdump-tools || true
+sudo sed -i 's/\/MODULES=dep\//\/MODULES=most\//' $FILESYSTEM_ROOT/etc/kernel/postinst.d/kdump-tools
+sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get install -y kdump-tools
```
Comment (qiluo-msft) [inline files/build_templates/sonic_debian_extension.j2]: Is it a workaround for some known bug? Should we add TODO comment after the known bug fixed?

File: files/build_templates/sonic_debian_extension.j2 [hunk 7]
```diff
@@ -353,6 +353,14 @@ sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get purge
...
+{% if sonic_kdump_enable == "y" %}
```
Comment (qiluo-msft) [inline files/build_templates/sonic_debian_extension.j2]: if-block should contain the upper 3 statements?

File: files/image_config/platform/rc.local [hunk 8]
```diff
@@ -319,6 +319,13 @@ if [ -f $FIRST_BOOT_FILE ]; then
         mv /host/grub.cfg /host/grub/grub.cfg
     fi
 
+    if `grep crashkernel=256M /proc/cmdline > /dev/null 2>&1`;
+    then
+	kdump-config unload
+	kdump-config symlinks `uname -r`
+	kdump-config load
+    fi
```
Comment (qiluo-msft) [inline files/image_config/platform/rc.local]: Follow code format nearby.

File: onie-mk-demo.sh [hunk 9]
```diff
@@ -89,6 +89,10 @@ output_raw_image=$(cat onie-image.conf | grep OUTPUT_RAW_IMAGE | cut -f2 -d"=")
 output_raw_image=$(eval echo $output_raw_image)
 
 # Tailor the demo installer for OS mode or DIAG mode
+if [ $sonic_kdump_enable = "y" ];
+then
+    sed -i "s/[^M] quiet/ crashkernel=256M quiet/" $tmp_installdir/install.sh
+fi
```
Comment (qiluo-msft) [inline onie-mk-demo.sh]: You can use the variable GRUB_CMDLINE_LINUX instead of directly modifying script

Comment (sun-siyuan) [reply onie-mk-demo.sh]: could you please elaborate on this suggestion?

**Comment by sun-siyuan** [comment]
vsimage failed due to no space error, please fix and trigger retest


WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Sending build context to Docker daemon   80.3MB

Step 1/18 : FROM debian:stretch
stretch: Pulling from library/debian
092586df9206: Pulling fs layer
092586df9206: Download complete
092586df9206: Pull complete
Digest: sha256:5fb93ce7a427b7c1c2374d5c29d68a159de7d5e781deeda422f8d51a1a9b6480
Status: Downloaded newer image for debian:stretch
 ---> cb15ecf641ad
Step 2/18 : RUN find /usr/share/doc -depth \( -type f -o -type l \) ! -name copyright | xargs rm || true
open /var/lib/docker/vfs/dir/3659c51f468f582034afd8032abac00a1480ab66ebeeaf2bfeef2030bc8e739a/usr/sbin/cppw: no space left on device
[  FAIL LOG END  ] [ target/docker-base-stretch.gz ]
slave.mk:526: recipe for target 'target/docker-base-stretch.gz' failed
make: *** [target/docker-base-stretch.gz] Error 1
Makefile.work:191: recipe for target 'target/sonic-vs.img.gz' failed
make[1]: *** [target/sonic-vs.img.gz] Error 2
make[1]: Leaving directory '/data/johnar/workspace/vs/buildimage-vs-image-pr@2'
Makefile:6: recipe for target 'target/sonic-vs.img.gz' failed
make: *** [target/sonic-vs.img.gz] Error 2

File: onie-mk-demo.sh (hunk 9 above)
Comment (sun-siyuan) [reply onie-mk-demo.sh]: @qiluo-msft 
could you please elaborate on this suggestion?

File: files/image_config/platform/rc.local (hunk 8 above)
Comment (qiluo-msft) [reply files/image_config/platform/rc.local]: I notice you mix space with tab. Please only use space in this file and follow code format nearby.

---
In reply to: [331242454](https://github.com/Azure/sonic-buildimage/pull/3276#discussion_r331242454) [](ancestors = 331242454)

File: onie-mk-demo.sh [hunk 10]
```diff
@@ -89,6 +89,10 @@ output_raw_image=$(cat onie-image.conf | grep OUTPUT_RAW_IMAGE | cut -f2 -d"=")
 output_raw_image=$(eval echo $output_raw_image)
 
 # Tailor the demo installer for OS mode or DIAG mode
+if [ $sonic_kdump_enable = "y" ];
+then
```
Comment (qiluo-msft) [inline onie-mk-demo.sh]: Please keep if-then in one line.

File: onie-mk-demo.sh (hunk 9 above)
Comment (qiluo-msft) [reply onie-mk-demo.sh]: Sorry I made a wrong suggestion. You may consider the macro EXTRA_CMDLINE_LINUX, but install.sh need add it at the same time. `sed` a bash script is not recommended.

---
In reply to: [331246069](https://github.com/Azure/sonic-buildimage/pull/3276#discussion_r331246069) [](ancestors = 331246069)

File: files/build_templates/sonic_debian_extension.j2 (hunk 7 above)
Comment (sun-siyuan) [reply files/build_templates/sonic_debian_extension.j2]: @qiluo-msft 

only install crash tool  when kdump_enable

Comment (qiluo-msft) [reply files/build_templates/sonic_debian_extension.j2]: I am wondering should we install kdump-tools and crash only when kdump_enable


---
In reply to: [338329974](https://github.com/Azure/sonic-buildimage/pull/3276#discussion_r338329974) [](ancestors = 338329974)

File: onie-mk-demo.sh (hunk 9 above)
Comment (sun-siyuan) [reply onie-mk-demo.sh]: @qiluo-msft 

there are big chunk of code sedding the bash script next to my change, I am doing in very similar way

sed -i -e "s/%%ENABLE_KDUMP%%/$sonic_kdump_enable/g" \
       -e "s/%%DEMO_TYPE%%/$demo_type/g" \
       -e "s/%%IMAGE_VERSION%%/$image_version/g" \
       -e "s/%%ONIE_IMAGE_PART_SIZE%%/$onie_image_part_size/" \
       -e "s/%%EXTRA_CMDLINE_LINUX%%/$EXTRA_CMDLINE_LINUX/" \
    $tmp_installdir/install.sh || clean_up 1

Comment (qiluo-msft) [reply onie-mk-demo.sh]: Please explorer this suggestion. The `sed` you mentioned is much conservative and safe.
> consider the macro EXTRA_CMDLINE_LINUX, but install.sh need add it at the same time

File: files/build_templates/sonic_debian_extension.j2 (hunk 7 above)
Comment (sun-siyuan) [reply files/build_templates/sonic_debian_extension.j2]: totally understood your wondering, but this is suggested by guohan, install kdump-tools by default but install crash tool when kdump_enabled

File: onie-mk-demo.sh (hunk 9 above)
Comment (sun-siyuan) [reply onie-mk-demo.sh]: @qiluo-msft 

please check my latest commit carefully, I am doing the same way as other sed but it still sed a script. EXTRA_CMDLINE_LINUX has been used already, I could not reuse it. if necessary , could we arrange a phone call to sort this out?

File: files/build_templates/sonic_debian_extension.j2 (hunk 7 above)
Comment (sun-siyuan) [reply files/build_templates/sonic_debian_extension.j2]: thanks a lot for the review and approval:)

File: onie-mk-demo.sh (hunk 9 above)
Comment (sun-siyuan) [reply onie-mk-demo.sh]: @qiluo-msft 

thanks a lot for the review and approval:)

**Comment by sun-siyuan** [comment]
test failed with below error, need retest

Setting status of 451664d5981b54b9e795b3b2a5dc25fb9f1d4bda to FAILURE with url https://sonic-jenkins.westus2.cloudapp.azure.com/job/broadcom/job/buildimage-brcm-all-pr/1278/ and message: 'Build finished. No test results found.'
Using context: broadcom
hudson.remoting.ProxyException: org.jenkinsci.plugins.workflow.steps.MissingContextVariableException: Required context class hudson.FilePath is missing
Perhaps you forgot to surround the code with a step that provides this, such as: node, dockerNode
	at org.jenkinsci.plugins.workflow.steps.StepDescriptor.checkContextAvailability(StepDescriptor.java:266)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:263)
Caused: hudson.remoting.ProxyException: org.codehaus.groovy.runtime.InvokerInvocationException: org.jenkinsci.plugins.workflow.steps.MissingContextVariableException: Required context class hudson.FilePath is missing
Perhaps you forgot to surround the code with a step that provides this, such as: node, dockerNode
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.replay(CpsStepContext.java:492)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:317)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeDescribable(DSL.java:417)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:182)

**Comment by sun-siyuan** [comment]
retest please

**Comment by sun-siyuan** [comment]
please retest

**Comment by sun-siyuan** [comment]
retest please

**Comment by sun-siyuan** [comment]
retest all please

**Comment by sun-siyuan** [comment]
retest this please

**Comment by lguohan** [comment]
core dump added by broadcom

**Pull request closed without merging**